### ISSUE TITLE
Lists can only accept positive numbers as indices

### ISSUE THREAD

**Issue by ian-otto** [xonsh/xonsh#4028]
## xonfig

<details>

```
$ xonfig
+------------------+-----------------+
| xonsh            | 0.9.24          |
| Python           | 3.9.0           |
| PLY              | 3.11            |
| have readline    | True            |
| prompt toolkit   | 3.0.8           |
| shell type       | prompt_toolkit  |
| pygments         | 2.7.2           |
| on posix         | True            |
| on linux         | True            |
| distro           | arch            |
| on darwin        | False           |
| on windows       | False           |
| on cygwin        | False           |
| on msys2         | False           |
| is superuser     | False           |
| default encoding | utf-8           |
| xonsh encoding   | utf-8           |
| encoding errors  | surrogateescape |
| on jupyter       | True            |
| jupyter kernel   | None            |
+------------------+-----------------+
```

</details>

## Expected Behavior
I expect to be able to pass code segments with negative array selectors or with statements without failure. For example:
```
$test_var = [1, 2, 3, 4]
$test_var[0]
# outputs 1
$test_var[-1]
# outputs 4
```

### Traceback (if applicable)

<details>

```
Traceback (most recent call last):
  File "/usr/lib/python3.9/site-packages/xonsh/environ.py", line 2243, in xonsh_script_run_control
    run_script_with_cache(filename, execer, ctx)
  File "/usr/lib/python3.9/site-packages/xonsh/codecache.py", line 160, in run_script_with_cache
    ccode = compile_code(filename, code, execer, glb, loc, mode)
  File "/usr/lib/python3.9/site-packages/xonsh/codecache.py", line 119, in compile_code
    ccode = execer.compile(code, glbs=glb, locs=loc, mode=mode, filename=filename)
  File "/usr/lib/python3.9/site-packages/xonsh/execer.py", line 135, in compile
    tree = self.parse(input, ctx, mode=mode, filename=filename, transform=transform)
  File "/usr/lib/python3.9/site-packages/xonsh/execer.py", line 95, in parse
    tree, input = self._parse_ctx_free(input, mode=mode, filename=filename)
  File "/usr/lib/python3.9/site-packages/xonsh/execer.py", line 219, in _parse_ctx_free
    tree = self.parser.parse(
  File "/usr/lib/python3.9/site-packages/xonsh/parsers/base.py", line 497, in parse
    tree = self.parser.parse(input=s, lexer=self.lexer, debug=debug_level)
  File "/usr/lib/python3.9/site-packages/xonsh/ply/ply/yacc.py", line 335, in parse
    return self.parseopt_notrack(input, lexer, debug, tracking, tokenfunc)
  File "/usr/lib/python3.9/site-packages/xonsh/ply/ply/yacc.py", line 1122, in parseopt_notrack
    p.callable(pslice)
  File "/usr/lib/python3.9/site-packages/xonsh/parsers/base.py", line 2221, in p_atom_expr
    p[0] = self.apply_trailers(p[1], p[2])
  File "/usr/lib/python3.9/site-packages/xonsh/parsers/base.py", line 2215, in apply_trailers
    assert False
AssertionError
```

</details>

## Steps to Reproduce
```
$test_var = [1, 2, 3, 4]
print($test_var[-1]) # dies from above traceback
print($test_var[len($test_var) - 1]) # dies from above traceback
print($test_var[len($test_var)]) # I would expect a KeyError, but we get the above traceback
```

## For community
⬇️  **Please click the 👍 reaction instead of leaving a `+1` or 👍  comment**

**Comment by gforsyth** [comment]
Hey @ian-otto -- thanks for reporting!  This works on my end, but I'm currently running `xonsh` under Python 3.7 -- I suspect that some of the AST changes in 3.9 are screwing with our parsing.  Can you see if running at <3.9 resolves this issue?  We'll add this to the list of things to fix with 3.9

**Comment by marcusatiliusregulus** [comment]
can confirm on archlinux with python 3.9:
```
$ l = [1, 2, 3, 4]
$ l[1]
2
$ l[-1]
xonsh: To log full traceback to a file set: $XONSH_TRACEBACK_LOGFILE = <filename>
Traceback (most recent call last):
  File "/usr/lib/python3.9/site-packages/xonsh/ptk_shell/shell.py", line 238, in _push
    code = self.execer.compile(src, mode="single", glbs=self.ctx, locs=None)
  File "/usr/lib/python3.9/site-packages/xonsh/__amalgam__.py", line 23992, in compile
    tree = self.parse(input, ctx, mode=mode, filename=filename, transform=transform)
  File "/usr/lib/python3.9/site-packages/xonsh/__amalgam__.py", line 23952, in parse
    tree, input = self._parse_ctx_free(input, mode=mode, filename=filename)
  File "/usr/lib/python3.9/site-packages/xonsh/__amalgam__.py", line 24076, in _parse_ctx_free
    tree = self.parser.parse(
  File "/usr/lib/python3.9/site-packages/xonsh/parsers/base.py", line 497, in parse
    tree = self.parser.parse(input=s, lexer=self.lexer, debug=debug_level)
  File "/usr/lib/python3.9/site-packages/xonsh/ply/ply/yacc.py", line 335, in parse
    return self.parseopt_notrack(input, lexer, debug, tracking, tokenfunc)
  File "/usr/lib/python3.9/site-packages/xonsh/ply/ply/yacc.py", line 1122, in parseopt_notrack
    p.callable(pslice)
  File "/usr/lib/python3.9/site-packages/xonsh/parsers/base.py", line 2221, in p_atom_expr
    p[0] = self.apply_trailers(p[1], p[2])
  File "/usr/lib/python3.9/site-packages/xonsh/parsers/base.py", line 2215, in apply_trailers
    assert False
AssertionError
```

**Comment by ian-otto** [comment]
@gforsyth Yep, that might be it. I think Arch just bumped the default Python version to 3.9 which relates to when the problem started to appear. Let me see about reverting my Python version to see if that helps.

**Comment by anki-code** [comment]
Fixed in master by #4101.


Tested in docker:
```python
docker run --rm -it python:3.9-slim /bin/bash -c "apt update && apt install -y git && pip install -U git+https://github.com/xonsh/xonsh.git && xonsh"
```
